Move model files from apps targets to WordPressData

mokagio opened the pull request:
Builds on top of https://github.com/wordpress-mobile/WordPress-iOS/pull/24374, with standalone commits from https://github.com/wordpress-mobile/WordPress-iOS/pull/24375 and https://github.com/wordpress-mobile/WordPress-iOS/pull/24376, part of #24165.

mokagio committed:
Configure `Package.swift` with WordPressData framework deps

mokagio committed:
Add `WordPress-Swift.h` to WordPressData to work around build fail

mokagio committed:
Add prefix header

mokagio committed:
Add all required sources to WordPressData

mokagio committed:
Add required imports to Swift layer

mokagio committed:
Add `Blog+SelfHosted` to WordPressData, including dependencies

mokagio committed:
Hack `ContextManager.swift` to compile

mokagio committed:
Extract `defaultWordPressComAccountRestAPI` method to WordPressData

Only the method, not the rest of the extension to avoid dependencies
galore.

mokagio committed:
Replicate Keystone preprocessor macro hack for WordPressData-Swift.h

mokagio committed:
Add `Media+Blog.swift` to WordPressData to compile `PostHelper.m`

mokagio committed:
Add `RemotePost+Metadata` source to WordPressData for `PostHelper.m`

mokagio committed:
Add `PostServiceRemoteFactory` to WordPressData to fix `PostService.m`

mokagio committed:
Add `PostHelper+JetpackSocial` to WordPressData to fix `PostHelper.m`

mokagio committed:
Add `WordPressOrgRestApi+WordPress` to WordPressData

Made possible by
https://github.com/wordpress-mobile/WordPress-iOS/pull/24351

mokagio committed:
Add `.m` file in WordPressData only to config CocoaLumberJack

mokagio committed:
Add some `import CocoaLumberJackSwift` required by WordPressData

dangermattic commented:
<!--
  0 Errors
  3 Warnings: <code>Modules/Package.swift</c...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Use `_private_wordPressComRestApi` in `Blog+WordPressComRestAPI.swift` to build WordPressData

mokagio commented on Sources/WordPressData/CocoaLumberJackConfiguration.h:
```diff
@@ -0,0 +1 @@
+// Here only to have the matching .m where to define the DDLogLevel value.
```
Might not even be necessary. Gotta try that...

mokagio commented on Sources/WordPressData/WordPress-Swift.h:
```diff
@@ -0,0 +1,4 @@
+#ifndef WordPress_Swift_h
+#define WordPress_Swift_h
+// FIXME: Defined here to avoid compilation errors while files are migrated over...
+#endif /* WordPress_Swift_h */
```
Should be able to remove this now...

mokagio commented on WordPress/Classes/Models/Blog/Blog+WordPressComRestAPI.swift:
```diff
@@ -11,7 +11,8 @@ public extension Blog {
     var wordPressComRestApi: WordPressComRestApi? {
-        account?.wordPressComRestApi
+        // FIXME: We are banking on the fact that by the time a consumer reads this, WPAccount will already have initialized its API client instance.
+        account?._private_wordPressComRestApi
```
Maybe we could decouple `WPAccount` `wordPressComRestApi` from WordPressAuthenticator using... a `Notification`? 

https://github.com/wordpress-mobile/WordPress-iOS/blob/b14619bc0426c23af1400a32bc470a416099e19e/WordPress/Classes/Models/WPAccount%2BRestApi.swift#L12-L20

https://github.com/wordpress-mobile/WordPress-iOS/blob/b14619bc0426c23af1400a32bc470a416099e19e/WordPress/Classes/Models/WPAccount%2BRestApi.swift#L27-L38

I suggest a `Notification` because I don't think injecting callbacks or using a delegate would be feasible with a model like this.

mokagio replied:
See https://github.com/wordpress-mobile/WordPress-iOS/pull/24392

mokagio committed:
Add CocoaLumberjack as a dependency to the extensions to fix compilation

Otherwise, we'd get:

```
Undefined symbols for architecture arm64:
  "_OBJC_CLASS_$_DDLog", referenced from:
       in AppExtensionsService.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

mokagio committed:
Remove `WordPress-Swift.h` hack from WordPressData

mokagio committed:
Add `Blog+Lookup` to WordPressData

mokagio committed:
Add `WordPress.xcdatamodeld` to WordPressData

mokagio committed:
Duplicate test infra in WordPressData in order to run `BlogTests`

mokagio committed:
Remove unnecessary `CocoaLumberJackConfiguration.h` header

We don't need the header, the implementation file is enough for the
CocoaLumberJack shenanigans to take place.

mokagio commented on WordPress/Classes/Models/Blog/Blog.m:
```diff
@@ -569,9 +572,13 @@ - (BOOL)supports:(BlogFeature)feature
         case BlogFeatureStats:
             return [self supportsRestApi] && [self isViewingStatsAllowed];
         case BlogFeatureStockPhotos:
-            return [self supportsRestApi] && [JetpackFeaturesRemovalCoordinator jetpackFeaturesEnabled];
+            // FIXME: See https://github.com/wordpress-mobile/WordPress-iOS/pull/24332
+//            return [self supportsRestApi] && [JetpackFeaturesRemovalCoordinator jetpackFeaturesEnabled];
+            return [self supportsRestApi];
         case BlogFeatureTenor:
-            return [JetpackFeaturesRemovalCoordinator jetpackFeaturesEnabled];
+            // FIXME: See https://github.com/wordpress-mobile/WordPress-iOS/pull/24332
+//            return [JetpackFeaturesRemovalCoordinator jetpackFeaturesEnabled];
+            return NO;
```
https://github.com/wordpress-mobile/WordPress-iOS/pull/24393 should address this.

mokagio commented on WordPress/Classes/Utility/CoreData/ContextManager.swift:
```diff
@@ -234,13 +235,14 @@ private extension ContextManager {
 
-        let startupEvent = SentryStartupEvent()
+        // FIXME: Import the Sentry stuff, too — But it accesses the app delegate!
+        // let startupEvent = SentryStartupEvent()
```
@crazytonyli is this `ContextManager` Sentry logging something that we need, or is it by any chance something that we added at the time of adoption just to monitor 🤞 ?

Being able to just chop it off would simplify things.

mokagio replied:
See also p1743115991824589/1743090023.463689-slack-C08HQR4C2TS

crazytonyli replied:
I'm sorry I missed this message. I don't think I added the events. It's very likely that they were there before my time. I doubt these events ever occur on production. I did a quick search on Sentry and don't see any related events. However, I think it's best to keep them in case we introduce some issues to the Core Data layer. At the bare minimum, we can send the events to the app via a notification.

mokagio committed:
Remove WordPressData types from WordPress and Jetpack

Not Keystone yet because I don't know if that compiles okay to verify
the change.

mokagio committed:
Update imports after removing WordPressData files from WP and Jetpack

mokagio committed:
Disambiguate `Notification` usages — `/: Notification\)/`

mokagio committed:
Disambiguate `Notification` usages — `/: Notification$/`

mokagio committed:
Update imports after `Notification` disambiguation

mokagio committed:
Disambiguate `Notification` usages — `/, Notification(?!s)/`

mokagio committed:
Disambiguate `Notification` usages — `/as\? Notification/`

mokagio committed:
Add CocoaLumberjack to notification service extensions

mokagio committed:
Disambiguate `Notification` usages — `/ Notification \{/`

mokagio committed:
Disambiguate `Notification` usages — `/ Notification\?/`

mokagio committed:
Replace all `#import` of headers now in WordPressData with `@import WordPressData`

mokagio committed:
Disambiguate `Notification` usages — `/ Notification,$/`

mokagio committed:
Disambiguate `Notification` usages — `/\[Notification\]/`

mokagio committed:
Disambiguate `Notification` usages — `/ Notification\:/`

mokagio committed:
Disambiguate `Notification` usages — `/: Notification,/`

mokagio committed:
Work around (temporarily?!) a build failure after the WordPressData move

mokagio committed:
Remove unused `AccountService` import from `Blog.m`

It matters because once `Blog` is in WordPressData, `AccountService`
will no longer be available.

mokagio committed:
Disambiguate `Notification` usages — `/\(Notification\)/`

mokagio committed:
Remove WordPressData types from `WordPress-Bridging-Header.h`

mokagio committed:
Move some `MenuItem` imports from `.h` to `.m`

`MenuItem` was not used in the interface. Moving the import in the
implementation will make it cleaner to update once WordPressData is
integrated.

mokagio committed:
Disambiguate `Notification` usages — `/Notification else/`

mokagio committed:
Disambiguate `Notification` usages — `/Notification\!/`

mokagio committed:
Disambiguate `Notification` usages — `/as\? Notification/`

mokagio committed:
Address module verifier warnings in WordPressData

mokagio committed:
Extract Swift-dependant logic from `Theme`

mokagio committed:
Make `Media` extension `public`

mokagio committed:
Make WordPressData Swift API available to ObjC framework consumers

mokagio committed:
Remove unused WordPressData import from NotificationServiceExtension

mokagio committed:
Add CocoaLumberjack to extensions Jetpack uses to avoid compile failure

Undefined symbol: _OBJC_CLASS_$_DDLog

mokagio committed:
Add `testable import WordPressData` to unit tests

mokagio committed:
Update `WordPress.Notification` to `WordPressData.Notification` in tests

mokagio committed:
Add more required `@testable import WordPressData`

Via

```
find WordPress/WordPressTest \
  -name "*.swift" \
  -type f \
  -exec perl -i -0pe 's/(@testable import WordPress\n)(\nclass (\w+): CoreDataTestCase \{)/\1\@testable import WordPressData\n\nclass \3: CoreDataTestCase {/g' \
  {} \;
```

and

```
find WordPress/WordPressTest \
  -name "*.swift" \
  -type f \
  -exec perl -i -0pe 's/(@testable import WordPress\n)(\nfinal class (\w+): CoreDataTestCase \{)/\1\@testable import WordPressData\n\nfinal class \3: CoreDataTestCase {/g' \
  {} \;
```

mokagio committed:
Add WordPressData as a WordPressTests dependency

mokagio committed:
Add missing dependencies to WordPressTesting post WordPressData addition

mokagio committed:
Switch all models to use "Global Namespace"

Via

```
sed -i '' -E 's/representedClassName="(WordPress)?(\.)?([^"]+)"/representedClassName=".\3"/g' \
  Sources/WordPressData/Resources/WordPress.xcdatamodeld/WordPress\ 155.xcdatamodel/contents
```

which switched to "Current Product Module" but resulted in unit tests
failure in WordPressData, followed by

```
➜ sed -i '' -E 's/representedClassName="\.([^"]+)"/representedClassName="\1"/g' \
  Sources/WordPressData/Resources/WordPress.xcdatamodeld/WordPress\ 155.xcdatamodel/contents
```

mokagio committed:
REVERT ME - Only run WordPress tests

mokagio committed:
Load model from the `ContextManager` bundle in the unit tests

mokagio committed:
Replace `AppEnvironment` with `WordPressComRestApi.apiBaseURL` in RestAPI

Same rationale as 09990b8ee7e33945c89e55463a889372a56478a2

mokagio committed:
No longer access `_private_wordPressComRestApi` in `Blog`

mokagio committed:
Fix WordPressData linkage in Keystone and move sources

mokagio committed:
Update access of types required by Keystone for WordPressData link

mokagio committed:
Explicitly add WebKit as a Keystone dependency

Triggered by a few errors like

```
In file included from /Users/gio/Developer/a8c/wpios/WordPress/Classes/ViewRelated/Suggestions/SuggestionsTableViewCell.m:3:
/Users/gio/Developer/a8c/wpios/DerivedData/WordPress/Build/Intermediates.noindex/WordPress.build/Debug-iphonesimulator/Keystone.build/DerivedSources/Keystone-Swift.h:295:9: fatal error: module 'WebKit' not found
  295 | @import WebKit;
      |  ~~~~~~~^~~~~~
```

which I got after linking WordPressData against Keystone and moving all
the sources there.

mokagio committed:
Revert `CookieJar` `public` access level

Apparently unnecessary, see
https://github.com/wordpress-mobile/WordPress-iOS/pull/24420#discussion_r2032150631

mokagio commented:
The current status here is that we are up to date with `trunk`, all targets build, but there are issues with the unit tests due to finding the models out of the framework.

I tried switching all to global namespace module, but it did not work [172031e](https://github.com/wordpress-mobile/WordPress-iOS/pull/24378/commits/172031e60cab70a72a37bfea8d86435b4097a669)

The situation is made a bit more complex by having a variety of ways with which we instantiate entities

<img width="972" alt="image" src="https://github.com/user-attachments/assets/8f30e5c6-8a87-4e2d-9490-5fdf57bc7ed4" />

Given there are only three possible core data module settings, one would expect it to be straightforward to try them and find the best one. Alas, it's proven a fiddly task for me. 

I'll spend a bit more time on this today in the hope of getting to the bottom of it before hading over to @kean who offered to take over the last bits of integration.

mokagio committed:
Replace all string values passed to `forEntityName` with `entityName()`

Done via

```
find . \
  \( -path './DerivedData' -o -path './.git' -o -path './vendor' \) \
  -prune -o -type f -name "*.swift" \
  -exec sed -i '' -E 's/forEntityName: "([A-Za-z_][A-Za-z0-9_]*)"/forEntityName: \1.entityName()/g' {} +
```

mokagio committed:
Replace all `NSStringFromClass` values passed to `forEntityName`

Done via

```
find . \
  \( -path './DerivedData' -o -path './.git' -o -path './vendor' \) \
  -prune -o -type f -name "*.swift" \
  -exec sed -i '' -E 's/forEntityName: NSStringFromClass\(([A-Za-z_][A-Za-z0-9_]*)\.self\)/forEntityName: \1.entityName()/g' {} +
```

The goal is to have a single way to get the entity name so we can adapt
it to the new WordPressData setup in one go.

mokagio committed:
Replace all `Self.classNameWithoutNamespaces()` with `entityName()`

The goal is to have a single way to get the entity name so we can adapt
it to the new WordPressData setup in one go.

mokagio committed:
Replace all `classNameWithoutNamespace()` calls with `entityName()`

Done via

```
find . \( -path './DerivedData' -o -path './.git' -o -path './vendor' \) \
  -prune -o -type f -name "*.swift" \
  -exec sed -i '' -E 's/forEntityName: ([A-Za-z_][A-Za-z0-9_]*)\.classNameWithoutNamespaces\(\)/forEntityName: \1.entityName()/g' {} +
```

The goal is to have a single way to get the entity name so we can adapt
it to the new WordPressData setup in one go.

kean committed:
Fix WordPressData runtime issues (#24494)

* Move remaining types to WordPressData

* Use Current Module for Swift-only types

* Fix ReaderTeamTopic compilation

* Add missing class names

* Remove unused SiteInfo

* Add Bundle.wordPressData and update DataMigratorTests

* Update CoreDataMigrationTests

* Fix unit tests (AppEnvironment required)

mokagio committed:
Add `BlobEntity` to WordPressData

mokagio commented on Sources/WordPressData/Swift/ContextManager.swift:
```diff
@@ -234,13 +234,14 @@ private extension ContextManager {
         }
 
-        let startupEvent = SentryStartupEvent()
+        // FIXME: Import the Sentry stuff, too — But it accesses the app delegate!
```
https://linear.app/a8c/issue/AINFRA-383

mokagio commented on Tests/WordPressDataTests/BlogTests.swift:
```diff
@@ -0,0 +1,310 @@
+        XCTAssertTrue(blog.supports(.pluginManagement))
+    }
+
+    // FIXME: Crashes because WPAccount fixture sets username and triggers BuildSettings access
+//    func testThatPluginManagementIsEnabledForJetpack() {
+//        let blog = BlogBuilder(mainContext)
+//            .withAnAccount()
+//            .withJetpack(version: "5.6", username: "test_user", email: "user@example.com")
+//            .with(isHostedAtWPCom: false)
+//            .with(isAdmin: true)
+//            .build()
+//
+//        XCTAssertTrue(blog.supports(.pluginManagement))
+//    }
```
https://linear.app/a8c/issue/AINFRA-384

mokagio commented on Tests/WordPressDataTests/Helpers/BlogBuilder.swift:
```diff
@@ -0,0 +1,245 @@
+/// Creates a Blog
+///
+/// Defaults to creating a self-hosted blog
+// FIXME: Currently duplicated from WordPressTests for ease of working. Eventually, should be extracted in a WordPressDataTestsHelper framework to share here and with the consumers.
```
https://linear.app/a8c/issue/AINFRA-385

mokagio committed:
Address straightforward `FIXME` notes

kean commented on Modules/Package.swift:
```diff
@@ -319,6 +341,18 @@ enum XcodeSupport {
+                // EmitSwiftModule normal arm64 (in target 'WordPressNotificationServiceExtension' from project 'WordPress')
+                //    cd /path/to/repo/WordPress
+                //
+                // <unknown>:0: error: missing required modules: 'CocoaLumberjack', 'CocoaLumberjackSwiftSupport'
```
(nit, let's address later) we now have [WPLogging.h](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/Modules/Sources/WordPressSharedObjC/include/WPLogging.h) for use in non-app targets (or any targets)

mokagio replied:
Noted. And tracked in https://linear.app/a8c/issue/AINFRA-424/adopt-wploggingh

kean commented on Modules/Package.swift:
```diff
@@ -307,6 +318,17 @@ enum XcodeSupport {
+                //
+                "BuildSettingsKit",
+                "FormattableContentKit",
+                "SFHFKeychainUtils",
```
This is duplicated dependency which produces a new warning.

```suggestion
```

mokagio replied:
I didn't see this comment, but noticed the warning elsewhere and addressed via https://github.com/wordpress-mobile/WordPress-iOS/pull/24514

kean commented on WordPress/WordPress.xcodeproj/project.pbxproj:
```diff
@@ -4414,14 +4432,20 @@
 				GCC_PREPROCESSOR_DEFINITIONS = (
 					"DEBUG=1",
 					"$(inherited)",
+					"WPDATA=1",
```
(nit) seems to be unused

mokagio replied:
Good catch! Thanks!

https://github.com/wordpress-mobile/WordPress-iOS/pull/24517
